src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,550 @@
+        private readonly IActiveConfiguredProjectSubscriptionService _projectSubscriptionService;
+        private readonly IFileSystem _fileSystem;
+        private readonly ITelemetryService _telemetryService;
+        private readonly IProjectFaultHandlerService _projectFaultHandlerService;
+
+        private readonly ITempPECompiler _compiler;
+        private readonly Lazy<VSBuildManager> _buildManager;
+
+        [ImportingConstructor]
+        public TempPEBuildManager(IProjectThreadingService threadingService,
+                                  IUnconfiguredProjectCommonServices unconfiguredProjectServices,
+                                  IActiveWorkspaceProjectContextHost activeWorkspaceProjectContextHost,
+                                  IActiveConfiguredProjectSubscriptionService projectSubscriptionService,
+                                  VSLangProj.VSProjectEvents projectEvents,
+                                  ITempPECompiler compiler,
+                                  IFileSystem fileSystem,
+                                  ITelemetryService telemetryService,
+                                  IProjectFaultHandlerService projectFaultHandlerService)
+             : base(threadingService.JoinableTaskContext)
+        {
+            UnconfiguredProjectServices = unconfiguredProjectServices;
+            _activeWorkspaceProjectContextHost = activeWorkspaceProjectContextHost;
+            _projectSubscriptionService = projectSubscriptionService;
+            _buildManager = new Lazy<VSBuildManager>(() => (VSBuildManager)projectEvents.BuildManagerEvents);
+            _compiler = compiler;
+            _fileSystem = fileSystem;
+            _telemetryService = telemetryService;
+            _projectFaultHandlerService = projectFaultHandlerService;
+        }
+
+        public string[] GetTempPEMonikers()
+        {
+            Initialize();
+
+            return AppliedValue.Value.Inputs.ToArray();
+        }
+
+        /// <summary>
+        /// Called privately to actually fire the TempPE events and optionally recompile the TempPE library for the specified input
+        /// </summary>
+        private async Task FireTempPEDirtyAsync(string moniker, bool shouldCompile)
+        {
+            // Not using use the ThreadingService property because unit tests
+            await UnconfiguredProjectServices.ThreadingService.SwitchToUIThread();
+
+            if (shouldCompile)
+            {
+                DesignTimeInputsItem inputs = AppliedValue.Value;
+                if (inputs.TaskSchedulers.TryGetValue(moniker, out ITaskDelayScheduler scheduler))
+                {
+                    HashSet<string> files = GetFilesToCompile(moniker, inputs.SharedInputs);
+                    string outputFileName = GetOutputFileName(inputs.OutputPath, moniker);
+                    _projectFaultHandlerService.Forget(scheduler.ScheduleAsyncTask(token => CompileTempPEAsync(files, outputFileName, token)).Task, UnconfiguredProject);
```

Comment:
Seems like there is a lot of scheduling logic both here and in `GetTempPEDescriptionXmlAsync`, perhaps this could be consolidated into a single helper method?